format/format.go

```diff
@@ -0,0 +1,406 @@
+package format
+
+import (
+	"cmp"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+	"slices"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/charmbracelet/log"
+	"github.com/gobwas/glob"
+	"github.com/numtide/treefmt/config"
+	"github.com/numtide/treefmt/stats"
+	"github.com/numtide/treefmt/walk"
+	"github.com/numtide/treefmt/walk/cache"
+	bolt "go.etcd.io/bbolt"
+	"golang.org/x/sync/errgroup"
+	"mvdan.cc/sh/v3/expand"
+)
+
+const (
+	batchKeySeparator = ":"
+)
+
+var ErrFormattingFailures = errors.New("formatting failures detected")
+
+// batchKey represents the unique sequence of formatters to be applied to a batch of files.
+// For example, "deadnix:statix:nixpkgs-fmt" indicates that deadnix should be applied first, statix second and
+// nixpkgs-fmt third.
+// Files are batched based on their formatting sequence, as determined by the priority and includes/excludes in the
+// formatter configuration.
+type batchKey string
+
+// sequence returns the list of formatters, by name, to be applied to a batch of files.
+func (b batchKey) sequence() []string {
+	return strings.Split(string(b), batchKeySeparator)
+}
+
+func newBatchKey(formatters []*Formatter) batchKey {
+	var key string
+
+	for _, f := range formatters {
+		key += f.name + batchKeySeparator
+	}
+
+	key = key[:len(key)-1]
```

Comment:
I don't remember why I did it like this. It might save on allocations but I don't think it will make much of a difference if I just switch it to join.

Reply:
Changed to use `strings.Join`